### DISCUSSION TITLE
High level chat interface & Ruby DSL for adding callable functions

### DISCUSSION THREAD

**Discussion by ErikDeBruijn** [alexrudall/ruby-openai#375] [Show and tell]
Firstly, let me say thanks for this gem @alexrudall !

On top op this gem I built the following for an intelligent agent:
1. High level interface for chat
2. Function registration and calling

TL;DR: I wanted to be able to do this:
```
agent = AI::Agent.new("You're an assistant. Today is November 10th, 2023.")
agent << "Can I get to Rotterdam in 3 hours?"
agent.chat!
# => (agent reponds about what modality)
agent << "Yes, by car"
# => (agent calls geo lookup functions and response is given, agent uses result to respond)
```

I am wondering whether some functionality could/should be integrated in the library or that a higher-level gem of mine should simply use this library. An interesting consideration is the fact that there is more than one LLM, and just like Faraday, it could be nice if there were Adapters that allow you to switch backends. I think registration of functions and calling them on the "ruby" side might be responsibilities of this library.

I can create a pull request for this, that will enable you to see how it's organized, but before I do this, I'd love to get your take on this.

### 1. High level interface for chat
Basically it enables this:
```
  describe "#chat!" do
    it "calls a single function and uses its result." do
      create(:account, email: "mail@myexample.nl", created_at: Time.zone.parse("2023-11-13"))

      VCR.use_cassette("open_ai/chat_function_call") do
        agent = AI::Agent.new("You're an assistant. Today is November 10th, 2023.")
        agent << "What is the account for mail@myexample.nl and how long ago was it created?"

        expect(agent.chat!).to eq(<<~RESPONSE.squish)
          The account associated with mail@myexample.nl was created on November 13th, 2023, which is 3 days from today.
        RESPONSE
      end
    end

    it "calls parallel function calls and uses their results." do
      create(:account, email: "test@stekker.com", created_at: Time.zone.parse("2023-11-13"))
      create(:account, email: "test2@stekker.com", created_at: Time.zone.parse("2023-11-10"))

      VCR.use_cassette("open_ai/chat_parallel_function_calls") do
        agent = AI::Agent.new(
          "You're an assistant. Today is November 10th, 2023.",
          model: "gpt-4-1106-preview",
        )
        agent << "Is the account test@stekker.com made before or after the account test2@stekker.com." \
                 "If so, by how many days?"

        expect(agent.chat!).to include("test@stekker.com was made 3 days after the account test2@stekker.com")
      end
    end
```

### 2. Function registration and calling
The function used above is declared as follows:
```
    declare_function(
      :customer_account,
      "Retrieve basic customer account details based on e-mail. Any further function calls will be " \
      "done with this account ID.",
      parameters: {
        account_email: {
          type: :string,
          description: "The e-mail address the user signed up his account with.",
        },
      },
    ) do |email_address|
      account = Account.find_by!(email: email_address)
      account.attributes
        .symbolize_keys
        .slice(:id, :name, :created_at)
    end
```
For this the agent advertises the available functions, has a `@messages` stack, calls the functions and returns the result and manages the conversation further.

I'm very interested in your thoughts!

**Comment by ErikDeBruijn** [comment]
@alexrudall did you see my post? Please just reply with a few thoughts, so I can start contributing something in one way or another.

**Reply by alexrudall** [reply]
Oh man, I never looked in Discussions before, sorry!! Let me take a look :D
